### DISCUSSION TITLE
Display variable

### DISCUSSION THREAD

**Discussion by wimgrymonprez** [forntoh/LcdMenu#67] [Q&A] [answered]
I'm trying to make a system for my brother, and i need a display to control everything.
I will be using 2 TDS sensors that measure the water purity. My brother is a window washer. So to have a proof of concept i'm creating the menu and i'm trying to create a submenu to display the sensor 'readings'. At the moment i'm just entering static text just to test if it diplays.

I've tried this : 

with an ItemInput i tried menu[1][2][1].value = "1"; in the loop function
with an MenuItem i tried menu[1][2][1].setText("1"); in the loop function

The strange thing is that the text displays, but then immediately after, the esp32 crashes.

This is my code:

#include <AiEsp32RotaryEncoder.h>
#include <LcdMenu.h>
#define LCD_ROWS 4
#define LCD_COLS 20

// setting up connection pins
// input controls
#define ROTARY_ENCODER_A_PIN 18
#define ROTARY_ENCODER_B_PIN 19
#define ROTARY_ENCODER_BUTTON_PIN 5
#define ROTARY_ENCODER_STEPS 3

// input sensors
#define TDS1 34
#define TDS2 35
#define FlowMeterIn 32
#define FlowMeterOut 33
#define WaterLevelTop 25
#define WaterLevelBottom 26

//output valves
#define MainValve 17
#define FlushValve 16
#define SaverValve1 4
#define SaverValve2 0
#define Power 2
String testvalue = "1.16";
// Stuff for rotary knob
AiEsp32RotaryEncoder rotaryEncoder = AiEsp32RotaryEncoder(ROTARY_ENCODER_A_PIN, ROTARY_ENCODER_B_PIN, ROTARY_ENCODER_BUTTON_PIN, -1, ROTARY_ENCODER_STEPS);
boolean up = false;
boolean down = false;
boolean middle = false;
char direction = 0;
int16_t last, value;

// Declare the call back function
void Backlight(uint8_t isOn);
void TDS1CALLBACK(String TDS1Value);
void goBack();

extern MenuItem mainMenu[];
extern MenuItem settingsMenu[];
extern MenuItem parameterInfo[];

MenuItem mainMenu[] = {ItemHeader(),
                       MenuItem("Start system"),
                       MenuItem("Connect to WiFi"),
                       ItemSubMenu("Check Parameters", parameterInfo),
                       ItemSubMenu("Settings", settingsMenu),
                       MenuItem("Flush system"),
                       MenuItem("Stop system"),
                       ItemFooter()};
/**
 * Create submenu and precise its parent
 */
MenuItem settingsMenu[] = {ItemHeader(mainMenu),
                           //
                           // Include callback in ItemToggle
                           //
                           ItemToggle("Backlight", Backlight),
                           ItemCommand("Back", goBack),
                           ItemFooter()};

/**
 * Create submenu and precise its parent
 */
MenuItem parameterInfo[] = {ItemHeader(mainMenu),
                           ItemInput("TDS1","0",TDS1CALLBACK),
                           ItemInput("TDS2","0",TDS1CALLBACK),
                           ItemCommand("Back", goBack),
                           ItemFooter()};

LcdMenu menu(LCD_ROWS, LCD_COLS);

void IRAM_ATTR readEncoderISR()
{
    rotaryEncoder.readEncoder_ISR();
}

void readRotaryEncoder()
{
  value = rotaryEncoder.readEncoder();
  
  if (value > last) {
    last = value;
    direction = 0;
    Serial.println("down");
    delay(150);
  }else   if (value < last) {
    last = value;
    direction = 1;
    Serial.println("UP");
    delay(150);
  }
}

void setup()
{
    Serial.begin(115200);
    pinMode(TDS1, INPUT);
    pinMode(TDS2, INPUT);
    pinMode(FlowMeterIn, INPUT);
    pinMode(FlowMeterOut, INPUT);
    pinMode(WaterLevelTop, INPUT);
    pinMode(WaterLevelBottom, INPUT);
    pinMode(MainValve, OUTPUT);
    pinMode(FlushValve, OUTPUT);
    pinMode(SaverValve1, OUTPUT);
    pinMode(SaverValve2, OUTPUT);
      
    menu.setupLcdWithMenu(0x27, mainMenu);
    rotaryEncoder.begin();
    rotaryEncoder.setup(readEncoderISR);
    rotaryEncoder.setBoundaries(-1000, 1000, false); //minValue, maxValue, circleValues true|false (when max go to min and vice versa)
    rotaryEncoder.setAcceleration(250);
    last = -9999;
}

void loop()
{
    //menu.updateTimer();
    menu[1][2][1].value = "1";
    if (rotaryEncoder.encoderChanged())
    {
        readRotaryEncoder();

        if (direction == 1)
          menu.up();
        else if (direction == 0)
          menu.down();
        
    }
    if (rotaryEncoder.isEncoderButtonClicked())
        {
          Serial.println("button pressed");
          menu.enter();
          
        }
    
}

/**
 * Define callback
 */
void Backlight(uint8_t isOn=0x1) {
  Serial.print(menu.getCursorPosition());
  if(isOn == 0x1){
    menu.lcd->setBacklight(0x0);
  }
  else{
    menu.lcd->setBacklight(0x1);
    }
}

void goBack(){
  menu.back();
}

void TDS1CALLBACK(String TDS1Value){
      
}

**Comment by wimgrymonprez** [comment]
This is the error that i'm getting

Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x400d343b  PS      : 0x00060630  A0      : 0x800d3488  A1      : 0x3ffb27a0  
A2      : 0x00000064  A3      : 0x00000001  A4      : 0x00000000  A5      : 0x00000000  
A6      : 0x3ffb8a28  A7      : 0x00000000  A8      : 0x80088be4  A9      : 0x3ffb2750  
A10     : 0x3ffc1dc8  A11     : 0x00000000  A12     : 0x00000000  A13     : 0x00000001  
A14     : 0x00060620  A15     : 0x00000001  SAR     : 0x00000005  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000073  LBEG    : 0x40084609  LEND    : 0x40084611  LCOUNT  : 0x00000027  


Backtrace:0x400d3438:0x3ffb27a00x400d3485:0x3ffb27c0 0x400d360a:0x3ffb27e0 0x400d1ef9:0x3ffb2800 0x400d36ed:0x3ffb2820

**Reply by forntoh** [reply]
It's not a good idea to change the value in the loop. The loop function is executed so many times depending on the speed of the processor, so the crash is possibly because you are doing too much on the loop function

**Reply by forntoh** [reply]
A possible solution is to use timers, and also update the Menu only when the value is different

**Reply by wimgrymonprez** [reply]
Thank you very much for your reply. I tried with an interval of 5 seconds, but when i access the menu where the values are updated every 5 seconds, the esp crashed. When i don't go into this menu, the esp32 keeps running. I'm really puzzled.
I also got this error message when the esp32 boots, or resets E (5732) gpio: gpio_set_level(226): GPIO output gpio_num error.
This error is displayed **only when i have the display connected**. When i disconnect the lcd display this error does not appear.

This is the piece of code that does the update in the interval

CurrentMillis = millis();
  if (CurrentMillis - PreviousMillis >= (Interval * 1000))
  {
    PreviousMillis = CurrentMillis;
    if(i==0){
      i=1;
      }
    else{
      i=0;
      }
    menu[1][2][1].value = "1";

**Reply by wimgrymonprez** [reply]
I did some more testing. I set up an interval of 15 seconds. This way i have ample time to go in the menu, see the result en go back before the value is updated. I generate a random number every interval. to see that the value is changed.
Now every time i go in i can see the changed number, and i go back. The script keeps running. Only when i remain in the same menu, and the interval triggers, that's when the scripts crashes. Am i missing something?

**Reply by forntoh** [reply]
> Thank you very much for your reply. I tried with an interval of 5 seconds, but when i access the menu where the values are updated every 5 seconds, the esp crashed. When i don't go into this menu, the esp32 keeps running. I'm really puzzled.
> 
> I also got this error message when the esp32 boots, or resets E (5732) gpio: gpio_set_level(226): GPIO output gpio_num error.
> 
> This error is displayed **only when i have the display connected**. When i disconnect the lcd display this error does not appear.
> 
> 
> 
> This is the piece of code that does the update in the interval
> 
> 
> 
> CurrentMillis = millis();
> 
>   if (CurrentMillis - PreviousMillis >= (Interval * 1000))
> 
>   {
> 
>     PreviousMillis = CurrentMillis;
> 
>     if(i==0){
> 
>       i=1;
> 
>       }
> 
>     else{
> 
>       i=0;
> 
>       }
> 
>     menu[1][2][1].value = "1";
> 
> 
> 
> 
> 
> 
> 
> 

Try updating the menu in the if block

**Reply by wimgrymonprez** [reply]
Hello again,

I tried your suggestion, but it didn't change anything. The crash still happens. My board is the esp32 from az delivery. Is it possible there is some kind of compatibility issue?

**Reply by forntoh** [reply]
I will do some tests this weekend, but I don't own an ESP. This library is not well optimised at memory usage so the esps small memory might as well be the cause. Try to use no submenus and maybe limit the items to 2 and observe of it still crashes.

**Reply by wimgrymonprez** [reply]
Hello again, I think i found an issue. When i do this menu[1][2][1].setText("test"); it puts the text test where it needs to be. I check this with the interval that the text is there. But when i'm inside the menu when the update is happening, the script crashes. BUT when i do menu[1][1].setText("test"); The script updates the value in whatever menu or submenu i'm at that time. The value is updated without a crash. So i think there is a problem with that whem i'm in the submenu, it can't find the menu because it seems to be relative to my current position in the menu, instead of the absolute location.

**Reply by forntoh** [reply]
Yeah, that's right
https://github.com/forntoh/LcdMenu/blob/8bf6dee0830df5f2d973ab933813f54e2bf1524b/src/LcdMenu.h#L759-L767
Uses the current menu (or submenu). You update an item only when it's displayed. Nevertheless, you can update a child menu without first to navigating to it.

**Reply by forntoh** [reply]
You can use the menu you created directly, no need to reference it from the LcdMenu. They're the same objects. I hope that you understand what I mean.

I'll add another example to the Docs with this scenario.

**Reply by forntoh** [reply, marked as answer]
> Hello again, I think i found an issue. When i do this menu[1][2][1].setText("test"); it puts the text test where it needs to be. I check this with the interval that the text is there. But when i'm inside the menu when the update is happening, the script crashes. BUT when i do menu[1][1].setText("test"); The script updates the value in whatever menu or submenu i'm at that time. The value is updated without a crash. So i think there is a problem with that whem i'm in the submenu, it can't find the menu because it seems to be relative to my current position in the menu, instead of the absolute location.

You are doing `menu[1][2][1].setText("test");`
Say the submenu in the second position of the main menu is called `settingsMenu` in your code, you can access it directly by doing `settingsMenu[1].setText("test");`

**Reply by wimgrymonprez** [reply]
Thank you so much for this explanation and for your time. Now the field is updated without a crash. 
The only thing i need to figure out now is how to make it update without me having to move the rotary encoder. Because the screen isn't updated when i'm not moving the encoder.
The last thing i'm struggling with is this:
I'm trying to accomplish this. A tds sensor reads the purity of the water and sends this value to a variable. I need the screen to update this value so that this value can be seen in the parameters menu. It would then show for example "TDS1: 1.25". I'm using an inputmenu item because this way i only have to update the number part. If i use a regular menuitem i would have to concatenate string and float and convert it into a char array. Using the input items is fine, but if i where to push the rotary button, i'm entering editing mode, which i'm not using. But then i can't exit out. So i'm thinking there is a better way but i am unaware of it.

**Reply by forntoh** [reply]
For now, you can call `menu.show()` right after updating the text, it will repaint the menu.

I understand what you want to accomplish, I can add a new type of item menu just for that. Or add a field called `isEditable` to `ItemInput`

**Reply by wimgrymonprez** [reply]
wow, thank you so much. It's working like a charm now. For the time being i'm checking on the buttonpress for the rotary encoder like this.
menu.enter();
      if (menu.isInEditMode()){
        menu.back();
        }
So now the menu is working like a charm, and very responsive. Thank you again for the help and explaining what i had to do. I have looked at many menu systems over the last few months, but yours it the best one out there :-)

**Reply by forntoh** [reply]
I'm happy that the library was useful to you 😊
Consider giving the repo a star so that it can reach even more people.

Also, there will be a minor release later today, it will allow you to use `menu.update()` instead of `menu.show()` as it wasn't meant for your use case.

**Reply by wimgrymonprez** [reply]
Hello Forntoh,

Today i noticed the library has changed quite a bit since the last time i was working on my project. I wish to adapt my script. But in the examples i can't find a way to update text to show the current value of a variable. I also want this to happen even without pressing a button. I want the main menu to show the value of 4 sensors, and then a setup menu. But can i use basic menu items and update the text, or should i use another type of menuitem. And how can i set the value? Thank you in advance for your help.

**Reply by forntoh** [reply]
This discussion might help you
https://github.com/forntoh/LcdMenu/discussions/9#discussioncomment-1219606

**Comment by wimgrymonprez** [comment]
Thank you for the quick reply. I can't get it to work though.

I have this menu item

   // Define the main menu

MAIN_MENU(
    ITEM_BASIC("OSMOSESYSTEM"),
    ITEM_INPUT("TDS1:",inputCallback),
);

LcdMenu menu(LCD_ROWS, LCD_COLS);

in the loop I do this

char buffer[12];
menu[1]->value = sprintf(buffer,"TDS1: %.2f",TDS_Inlet);

i get this error 'class MenuItem' has no member named 'value'

I'm clearly doing something wrong, but it don't know what it is.

Verzonden vanuit Outlook<http://aka.ms/weboutlook>
________________________________
Van: Thomas Forntoh ***@***.***>
Verzonden: woensdag 9 augustus 2023 14:07
Aan: forntoh/LcdMenu ***@***.***>
CC: Wim Grymonprez ***@***.***>; Author ***@***.***>
Onderwerp: Re: [forntoh/LcdMenu] Display variable (Discussion #67)


This discussion might help you
#9 (comment)<https://github.com/forntoh/LcdMenu/discussions/9#discussioncomment-1219606>

**Reply by forntoh** [reply]
`menu[1]->setValue("AAAAAA")`
value is private, you have to use the setter.

**Comment by wimgrymonprez** [comment]
Brilliant, it works. I noticed I could have found this in the src directory. I'll do this first should i encounter another issue. Thank you very much for your help.

Verzonden vanuit Outlook<http://aka.ms/weboutlook>
________________________________
Van: Thomas Forntoh ***@***.***>
Verzonden: woensdag 9 augustus 2023 16:10
Aan: forntoh/LcdMenu ***@***.***>
CC: Wim Grymonprez ***@***.***>; Author ***@***.***>
Onderwerp: Re: [forntoh/LcdMenu] Display variable (Discussion #67)


menu[1]->setValue("AAAAAA")
value is private, you have to use the setter.

**Comment by wimgrymonprez** [comment]
Dear Forntoh,

I'm sorry to contact you again. I thought the menu worked, but when i got back from work i uploaded it to my esp32 and although it doesn't show an error, the value isn't being updated on the screen.

Here is the relevant code

MAIN_MENU(
    ITEM_BASIC("OSMOSERINATOR"),
    ITEM_INPUT("TDS1:",inputCallback),
    ITEM_SUBMENU("Control", onOffAuto)
);

In the loop I send these commands to the lcd

      menu[1]->setValue("TDS1: testtest");
      menu.update();

i also tried setText. But i think setText is for basic menu items?

I've been trying a lot of different things to no avail.

Verzonden vanuit Outlook<http://aka.ms/weboutlook>
________________________________
Van: Thomas Forntoh ***@***.***>
Verzonden: woensdag 9 augustus 2023 16:10
Aan: forntoh/LcdMenu ***@***.***>
CC: Wim Grymonprez ***@***.***>; Author ***@***.***>
Onderwerp: Re: [forntoh/LcdMenu] Display variable (Discussion #67)


menu[1]->setValue("AAAAAA")
value is private, you have to use the setter.

**Reply by forntoh** [reply]
Share the complete code (create a gist)

**Reply by forntoh** [reply]
`menu[2]` for `ITEM_INPUT("TDS1:",inputCallback)`

The item indexes start from 1

MAIN_MENU(
1:    ITEM_BASIC("OSMOSERINATOR"),
2:    ITEM_INPUT("TDS1:",inputCallback),
3:    ITEM_SUBMENU("Control", onOffAuto)
);

**Comment by wimgrymonprez** [comment]
Hello Thomas,

I've put the sketch on github and i've sent you an invite. This is the whole code, but it is a little long because of all the other things the script does. It also works with the arduino cloud dashboard. I just would like to also use the lcd for local control without having to use my cellphone.

In case you are interested in what this project is for:
This is a project for my brother. He is a window cleaner and he needs pure water for cleaning the windows so he doesn't need to dry after cleaning. When the pure water has dried it doesn't leave a stain. I'm still testing it, so not all variables in real minutes although it says so in the comments i put after.


Verzonden vanuit Outlook<http://aka.ms/weboutlook>
________________________________
Van: Thomas Forntoh ***@***.***>
Verzonden: woensdag 9 augustus 2023 23:42
Aan: forntoh/LcdMenu ***@***.***>
CC: Wim Grymonprez ***@***.***>; Author ***@***.***>
Onderwerp: Re: [forntoh/LcdMenu] Display variable (Discussion #67)


menu[2] for ITEM_INPUT("TDS1:",inputCallback),

The item indexes start from 1

MAIN_MENU(

  1.  ITEM_BASIC("OSMOSERINATOR"),
  2.  ITEM_INPUT("TDS1:",inputCallback),
  3.  ITEM_SUBMENU("Control", onOffAuto)
);
